wellcomeml/ml/bert_semantic_equivalence.py

```diff
@@ -86,32 +93,48 @@ def initialise_models(self):
         )
         return self.model
 
-    def _prep_dataset_generator(self, X, y):
+    def _prep_dataset_generator(self, X, y=None, shuffle=False, repeat=False, batch_size=32):
         features = ["input_ids", "attention_mask", "token_type_ids"]
 
+        if y is None:
+            input_element_types = {feature: tf.int32 for feature in features}
+            input_element_tensors = (
+                {feature: tf.TensorShape([None]) for feature in features}
+            )
+
+        else:
+            input_element_types = ({feature: tf.int32 for feature in features}, tf.int64)
+            input_element_tensors = (
+                {feature: tf.TensorShape([None]) for feature in features},
+                tf.TensorShape([]),
+            )
+
         batch_encoding = self.tokenizer.batch_encode_plus(
             X, max_length=self.max_length, add_special_tokens=True,
         )
 
         def gen_train():
             for i in range(len(X)):
-                features = {
+                features_dict = {
                     k: pad(batch_encoding[k][i], self.max_length)
                     for k in batch_encoding
                 }
-
-                yield (features, int(y[i]))
-
-        input_element_types = ({feature: tf.int32 for feature in features}, tf.int64)
-        input_element_tensors = (
-            {feature: tf.TensorShape([None]) for feature in features},
-            tf.TensorShape([]),
-        )
+                if y is None:
+                    yield features_dict
+                else:
+                    yield (features_dict, int(y[i]))
 
         dataset = tf.data.Dataset.from_generator(
             gen_train, input_element_types, input_element_tensors,
         )
 
+        if shuffle:
+            dataset = dataset.shuffle(self.buffer_size, seed=self.random_seed)
+
+        dataset = dataset.batch(batch_size)
+        if repeat:
+            dataset = dataset.repeat(-1)
```

Comment:
btw you do not need to repeat, if you do not define steps_per_epoch it figures it out on its own